ykrt/src/compile/jitc_yk/codegen/x86_64.rs

```diff
@@ -64,34 +64,80 @@ extern "C" fn __yk_deopt(frameaddr: *const c_void, deoptid: usize, jitrbp: *cons
     let newstack = unsafe { libc::malloc(memsize) };
     let mut rsp = unsafe { newstack.byte_add(memsize) };
+    let mut rbp = rsp;
+    // Keep track of the real address of the current frame so we can write pushed RBP values.
+    let mut lastframeaddr = frameaddr;
+    let mut lastframesize = 0;
 
     // Live register values that we need to write back into AOT registers.
     let mut registers = [0; 16];
-    let mut newframedst = frameaddr;
     let mut varidx = 0;
     for (frameid, smid) in info.frames.iter().enumerate() {
         let (rec, pinfo) = aot_smaps.get(usize::try_from(*smid).unwrap());
 
-        if frameid == 0 {
-            // Compute the address to which we want to write the new stack. This is immediately
-            // after the frame containing the control point.
-            newframedst = unsafe { newframedst.byte_sub(usize::try_from(rec.size).unwrap()) };
-            if pinfo.hasfp {
-                newframedst = unsafe { newframedst.byte_add(REG64_SIZE) };
+        // WRITE RBP
+        // If the current frame has pushed RBP we need to do the same (unless we are processing
+        // the bottom-most frame).
+        if pinfo.hasfp && frameid > 0 {
+            rsp = unsafe { rsp.sub(REG64_SIZE) };
+            rbp = rsp;
+            unsafe { ptr::write(rsp as *mut u64, lastframeaddr as u64) };
+        }
+
+        // Calculate the this frame's address by substracting the last frame's size (plus return
+        // address) from the last frame's address.
+        lastframeaddr = unsafe { lastframeaddr.byte_sub(lastframesize + REG64_SIZE) };
+        lastframesize = usize::try_from(rec.size).unwrap();
+
+        // Update RBP to represent this frame's address.
+        if pinfo.hasfp {
+            registers[usize::from(RBP_DWARF_NUM)] = lastframeaddr as u64;
+        }
+
+        // Now we write any callee-saved registers onto the new stack. Note, that if we have
+        // pushed RBP above (which includes adjusting RBP) we need to temporarily re-adjust our
+        // pointer. This is because the CSR index calculates from the bottom of the frame, not
+        // from RBP. For example, a typical prologue looks like this:
+        //   push rbp
+        //   mov rbp, rsp
+        //   push rbx     # this has index -2
+        //   push r14     # this has index -3
+        if frameid > 0 {
+            for (reg, idx) in &pinfo.csrs {
+                let mut tmp =
+                    unsafe { rbp.byte_sub(usize::try_from(idx.abs()).unwrap() * REG64_SIZE) };
+                if pinfo.hasfp {
+                    tmp = unsafe { tmp.byte_add(REG64_SIZE) };
+                }
+                let val = registers[usize::from(*reg)];
+                unsafe { ptr::write(tmp as *mut u64, val) };
             }
         }
 
+        // Now write all live variables to the new stack in the order they are listed in the AOT
+        // stackmap.
         for aotvar in rec.live_vars.iter() {
```

Comment:
I *think* varidx can be changed to `for (varidx, aotvar) in rec.live_vars.iter().enumerate()` and we can remove the definition at the top of the function and `varidx += 1` below?

Reply:
Sadly not. We want to increment `varidx` for each variable in each frame. Look where `varidx` is defined: https://github.com/ykjit/yk/pull/1096/files/28eaa49b399e203459f7d0c35c46853b155b83b4#diff-5d16a3250e41186bccaf33f5bf5d3c0a368044a529f1ee60958ee80a93531686R96

I'm not a fan of incrementing `varidx` like this either, and it already lead to a bug earlier because I forgot about a `continue`, which is why it was moved further up. If you have a better idea I'm all ears.